activerecord/lib/active_record/reflection.rb

```diff
@@ -700,10 +701,25 @@ def inverse_name
         def automatic_inverse_of
           if can_find_inverse_of_automatically?(self)
             inverse_name = ActiveSupport::Inflector.underscore(options[:as] || active_record.name.demodulize).to_sym
-            plural_inverse_name = ActiveSupport::Inflector.pluralize(inverse_name)
 
             begin
-              reflection = klass._reflect_on_association(inverse_name) || klass._reflect_on_association(plural_inverse_name)
+              reflection = klass._reflect_on_association(inverse_name)
+              if !reflection
+                plural_inverse_name = ActiveSupport::Inflector.pluralize(inverse_name)
+                reflection = klass._reflect_on_association(plural_inverse_name)
+
+                unless active_record.automatically_invert_plural_associations
+                  ActiveRecord.deprecator.warn(
+                    "The #{active_record.name}.#{name} inverse association could have been automatically" \
+                    " inferred as #{klass.name}.#{plural_inverse_name} but wasn't because `automatically_invert_plural_associations`" \
+                    " is disabled.\n\n" \
+                    "If this automatic inferrence is correct you can consider enabling" \
+                    " `config.active_record.automatically_invert_plural_associations`.\n\n" \
+                    "If it is incorrect, you can silence this warning by passing `inverse_of: false`."
```

Comment:
Similar to https://github.com/rails/rails/pull/50883/files#r1534290383, what do you think about shifting toward intent?

```suggestion
                    "If automatic inference is not intended, you can silence this warning by defining the association with `inverse_of: false`."
```